File: volozhients/compiler/src/main/java/graphlang/semantics/Scope.java
package graphlang.semantics;

import graphlang.semantics.symbols.Symbol;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Optional;

public final class Scope {

    private final Scope parent;
    private final Map<String, Symbol> symbols = new LinkedHashMap<>();

    public Scope(Scope parent) {
        this.parent = parent;
    }

    public Scope parent() {
        return parent;
    }

    public void define(Symbol symbol) {
        if (symbols.containsKey(symbol.name())) {
            throw new SemanticException("Symbol '%s' already declared in this scope".formatted(symbol.name()));
        }
        symbols.put(symbol.name(), symbol);
    }

    public Optional<Symbol> resolve(String name) {
        Scope scope = this;
        while (scope != null) {
            Symbol symbol = scope.symbols.get(name);
            if (symbol != null) {
                return Optional.of(symbol);
            }
            scope = scope.parent;
        }
        return Optional.empty();
    }

    public Optional<Symbol> resolveLocal(String name) {
        return Optional.ofNullable(symbols.get(name));
    }
}

File: demenchuk/compiler/semantic/scope.py
"""
Класс Scope для управления областями видимости
"""

from typing import Dict, List, Optional
from .symbol import Symbol


class Scope:
    """Область видимости"""
    
    def __init__(self, name: str = "global", parent: Optional['Scope'] = None):
        self.name = name
        self.parent = parent
        self.symbols: Dict[str, Symbol] = {}
        self.children: List['Scope'] = []
        
        if parent:
            parent.children.append(self)
    
    def define(self, symbol: Symbol) -> bool:
        """Добавляет символ в текущую область"""
        if symbol.name in self.symbols:
            return False  # Уже определён
        self.symbols[symbol.name] = symbol
        return True
    
    def lookup(self, name: str) -> Optional[Symbol]:
        """Ищет символ в текущей области и родительских"""
        if name in self.symbols:
            return self.symbols[name]
        if self.parent:
            return self.parent.lookup(name)
        return None
    
    def lookup_local(self, name: str) -> Optional[Symbol]:
        """Ищет символ только в текущей области"""
        return self.symbols.get(name)
    
    def __repr__(self):
        return f"Scope({self.name}, {len(self.symbols)} symbols)"
